rust/hyperlane-core/src/chain.rs

```diff
@@ -289,30 +242,42 @@ impl Debug for HyperlaneDomain {
 }
 
 impl HyperlaneDomain {
-    pub fn from_config(domain_id: u32, name: &str) -> Result<Self, &'static str> {
+    pub fn from_config(
+        domain_id: u32,
+        name: &str,
+        implementation: HyperlaneDomainImpl,
+    ) -> Result<Self, &'static str> {
+        let name = name.to_ascii_lowercase();
         if let Ok(domain) = KnownHyperlaneDomain::try_from(domain_id) {
-            if name.to_ascii_lowercase().as_str() == domain.as_str() {
+            if name == domain.as_str() {
                 Ok(HyperlaneDomain::Known(domain))
             } else {
                 Err("Chain name does not match the name of a known domain id; the config is probably wrong.")
             }
+        } else if name.as_str().parse::<KnownHyperlaneDomain>().is_ok() {
+            Err("Chain name is known the domain is incorrect; the config is probably wrong.")
         } else {
             Ok(HyperlaneDomain::Unknown {
                 domain_id,
-                chain_name: name.to_owned(),
+                chain_name: name,
                 // we might want to support accepting these from the config later
                 domain_type: HyperlaneDomainType::Unknown,
-                domain_impl: HyperlaneDomainImpl::Unknown,
+                domain_impl: implementation,
             })
         }
     }
 
-    pub fn from_config_strs(domain_id: &str, name: &str) -> Result<Self, &'static str> {
+    pub fn from_config_strs(
+        domain_id: &str,
+        name: &str,
+        implementation: HyperlaneDomainImpl,
```

Comment:
I don't feel too strongly, but I keep finding `implementation` a little unnatural / confusing. I think it's because my first reaction is to think of literal smart contract code or a smart contract address. Wondering how you feel about `protocol` instead?

Reply:
I am down for that.